bsonproto/bsonproto.go

```diff
@@ -0,0 +1,107 @@
+// Package bsonproto provides primitives for encoding and decoding of BSON.
+package bsonproto
+
+import (
+	"errors"
+	"fmt"
+	"time"
+)
+
+// ScalarType represents a BSON scalar type.
+type ScalarType interface {
+	float64 | string | Binary | ObjectID | bool | time.Time | NullType | Regex | int32 | Timestamp | int64
+}
+
+// Size returns a size of the encoding of value v in bytes.
+func Size[T ScalarType](v T) int {
+	return SizeAny(v)
+}
+
+// SizeAny returns a size of the encoding of value v in bytes.
+//
+// It panics if v is not a ScalarType.
+func SizeAny(v any) int {
+	switch v := any(v).(type) {
+	case float64:
+		return SizeFloat64(v)
+	case string:
+		return SizeString(v)
+	case int32:
+		return SizeInt32(v)
+	case int64:
+		return SizeInt64(v)
+	default:
+		panic(fmt.Sprintf("unsupported type %T", v))
+	}
+}
+
+// Encode encodes value v into b.
+//
+// b must be at least Size(v) bytes long; otherwise, Encode will panic.
+// Only b[0:Size(v)] bytes are modified.
+func Encode[T ScalarType](b []byte, v T) {
```

Comment:
Because I don't want those functions to allocate slices themselves. The caller might use some effective strategies like pools to make things faster. The design is closer to https://pkg.go.dev/encoding/binary#ByteOrder

Reply:
> Because I don't want those functions to grow slices themselves

Why? it will not grow if there is no need for that. As a result we don't panic on marshal (growing slice isn't a critical error).

Reply:
The whole point of this package is not to allocate memory itself and let the caller do that. `append` hides it from the caller.

Using `append` also would code encoding function to use slice capacity while decoding uses slice length. 

I think panic is justifiable there as it is a programmer's error – they are responsible for passing the large enough buffer, as returned by size functions.